File: framec_tests/golang/branch/branch_test.go
package branch

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func (m *branchStruct) log(msg string) {
	m.tape = append(m.tape, msg)
}

func TestSimpleIfBool(t *testing.T) {

	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.A()
	sm.OnBool(true)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"then 1", "then 2"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.A()
	sm.OnBool(false)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"else 1", "else 2"}, x.tape)
}

func TestSimpleIfInit(t *testing.T) {

	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.A()
	sm.OnInt(7)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"> 5", "< 10", "== 7"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.A()
	sm.OnInt(-3)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"<= 5", "< 10", "!= 7"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.A()
	sm.OnInt(12)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"> 5", ">= 10", "!= 7"}, x.tape)

}

func TestNegatedIfBool(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.B()
	sm.OnBool(true)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"else 1", "else 2"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.B()
	sm.OnBool(false)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"then 1", "then 2"}, x.tape)
}

func TestNegatedIfInt(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.B()
	sm.OnInt(7)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{">= 5", "<= 10", "== 7"}, x.tape)

	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.B()
	sm.OnInt(5)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{">= 5", "<= 10", "!= 7"}, x.tape)

	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.B()
	sm.OnInt(10)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{">= 5", "<= 10", "!= 7"}, x.tape)

	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.B()
	sm.OnInt(0)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"< 5", "<= 10", "!= 7"}, x.tape)

	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.B()
	sm.OnInt(100)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{">= 5", "> 10", "!= 7"}, x.tape)

}

func TestOperatorPrecedence(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.C()
	sm.OnInt(0)
	assert.Equal(t, []string{"then 1", "else 2", "then 3", "then 4"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(7)
	assert.Equal(t, []string{"else 1", "then 2", "else 3", "then 4"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(-3)
	assert.Equal(t, []string{"then 1", "else 2", "else 3", "else 4"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(12)
	assert.Equal(t, []string{"else 1", "else 2", "then 3", "else 4"}, x.tape)
}

func TestNestedIf(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.D()
	sm.OnInt(50)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"> 0", "< 100"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.D()
	sm.OnInt(200)
	assert.Equal(t, BranchState_NestedIf, x._compartment_.State)
	assert.Equal(t, []string{"> 0", ">= 100"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.D()
	sm.OnInt(-5)
	assert.Equal(t, BranchState_NestedIf, x._compartment_.State)
	assert.Equal(t, []string{"<= 0", "> -10"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.D()
	sm.OnInt(-10)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"<= 0", "<= -10"}, x.tape)
}

/// Test that a transition guarded by a conditional expression triggers an
/// early return from the handler.
func TestGuardedTransition(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.E()
	sm.OnInt(5)
	assert.Equal(t, BranchState_F3, x._compartment_.State)
	assert.Equal(t, []string{"-> $F3"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.E()
	sm.OnInt(15)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"-> $F2"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.E()
	sm.OnInt(115)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"-> $F1"}, x.tape)
}

/*  Test that a transition guarded by a nested conditional expression
triggers an early return from the handler, but this return doesn't
 apply to non-transitioned branches. */
func TestNestedGuardedTransition(t *testing.T) {
	sm := NewBranch()
	x := sm.(*branchStruct)
	sm.F()
	sm.OnInt(5)
	assert.Equal(t, BranchState_F3, x._compartment_.State)
	assert.Equal(t, []string{"-> $F3"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.F()
	sm.OnInt(15)
	assert.Equal(t, BranchState_F2, x._compartment_.State)
	assert.Equal(t, []string{"-> $F2"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.F()
	sm.OnInt(65)
	assert.Equal(t, BranchState_F3, x._compartment_.State)
	assert.Equal(t, []string{"-> $F3"}, x.tape)
	sm = NewBranch()
	x = sm.(*branchStruct)
	sm.F()
	sm.OnInt(115)
	assert.Equal(t, BranchState_F1, x._compartment_.State)
	assert.Equal(t, []string{"-> $F1"}, x.tape)
}

File: framec_tests/csharp/Branch/BranchTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using NUnit.Framework;
namespace Branch
{
    class BranchController : Branch
    {
        public BranchController() : base()
        {
        }
        public new void log_do(string msg)
        {
            this.tape.Add(msg);
        }
    }

    [TestFixture]
    public class BranchControllerTests
    {
        [Test]
        public void TestSimpleIfBool()
        {
            BranchController sm = new BranchController();
            sm.A();
            sm.OnBool(true);

            Assert.AreEqual("7", sm.state_info());
            Assert.AreEqual(new List<string> { "then 1", "then 2" }, sm.tape);

            sm = new BranchController();
            sm.A();
            sm.OnBool(false);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { "else 1", "else 2" }, sm.tape);
        }

        [Test]
        public void TestSimpleIfInt()
        {
            BranchController sm = new BranchController();
            sm.A();
            sm.OnInt(7);

            Assert.AreEqual("7", sm.state_info());
            Assert.AreEqual(new List<string> { "> 5", "< 10", "== 7" }, sm.tape);

            sm = new BranchController();
            sm.A();
            sm.OnInt(-3);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { "<= 5", "< 10", "!= 7" }, sm.tape);

            sm = new BranchController();
            sm.A();
            sm.OnInt(12);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { "> 5", ">= 10", "!= 7" }, sm.tape);
        }

        [Test]
        public void TestNegatedIfBool()
        {
            BranchController sm = new BranchController();
            sm.B();
            sm.OnBool(true);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { "else 1", "else 2" }, sm.tape);

            sm = new BranchController();
            sm.B();
            sm.OnBool(false);

            Assert.AreEqual("7", sm.state_info());
            Assert.AreEqual(new List<string> { "then 1", "then 2" }, sm.tape);
        }

        [Test]
        public void TestNegatedIfInt()
        {
            BranchController sm = new BranchController();
            sm.B();
            sm.OnInt(7);

            Assert.AreEqual("7", sm.state_info());
            Assert.AreEqual(new List<string> { ">= 5", "<= 10", "== 7" }, sm.tape);

            sm = new BranchController();
            sm.B();
            sm.OnInt(5);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { ">= 5", "<= 10", "!= 7" }, sm.tape);

            sm = new BranchController();
            sm.B();
            sm.OnInt(10);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string> { ">= 5", "<= 10", "!= 7" }, sm.tape);

            sm = new BranchController();
            sm.B();
            sm.OnInt(0);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string>() { "< 5", "<= 10", "!= 7" }, sm.tape);

            sm = new BranchController();

            sm.B();
            sm.OnInt(100);

            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(new List<string>() { ">= 5", "> 10", "!= 7" }, sm.tape);
        }

        [Test]
        public void OperatorPrecedenceTest()
        {
            BranchController sm = new BranchController();

            sm.C();
            sm.OnInt(0);
            Assert.AreEqual(new List<string>() { "then 1", "else 2", "then 3", "then 4" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(7);
            Assert.AreEqual(new List<string>() { "else 1", "then 2", "else 3", "then 4" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(-3);
            Assert.AreEqual(new List<string>() { "then 1", "else 2", "else 3", "else 4" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(12);
            Assert.AreEqual(new List<string>() { "else 1", "else 2", "then 3", "else 4" }, sm.tape);
        }

        [Test]
        public void NestedIfTest()
        {
            BranchController sm = new BranchController();

            sm.D();
            sm.OnInt(50);
            Assert.AreEqual("7", sm.state_info());
            Assert.AreEqual(new List<string>() { "> 0", "< 100" }, sm.tape);

            sm = new BranchController();
            sm.D();
            sm.OnInt(200);
            Assert.AreEqual("4", sm.state_info());
            Assert.AreEqual(new List<string>() { "> 0", ">= 100" }, sm.tape);

            sm = new BranchController();
            sm.D();
            sm.OnInt(-5);
            Assert.AreEqual("4", sm.state_info());
            Assert.AreEqual(new List<string>() { "<= 0", "> -10" }, sm.tape);

            sm = new BranchController();
            sm.D();
            sm.OnInt(-10);
            Assert.AreEqual("8", sm.state_info());
            Assert.AreEqual(sm.tape, new List<string> { "<= 0", "<= -10" });
        }

        [Test]
        public void TestGuardedTransition()
        {
            BranchController sm = new BranchController();
            sm.E();
            sm.OnInt(5);
            Assert.AreEqual(sm.state_info(), "9");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F3" });

            sm = new BranchController();
            sm.E();
            sm.OnInt(15);
            Assert.AreEqual(sm.state_info(), "8");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F2" });

            sm = new BranchController();
            sm.E();
            sm.OnInt(115);
            Assert.AreEqual(sm.state_info(), "7");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F1" });
        }

        [Test]
        public void TestNestedGuardedTransition()
        {
            BranchController sm = new BranchController();
            sm.F();
            sm.OnInt(5);
            Assert.AreEqual(sm.state_info(), "9");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F3" });

            sm = new BranchController();
            sm.F();
            sm.OnInt(15);
            Assert.AreEqual(sm.state_info(), "8");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F2" });

            sm = new BranchController();
            sm.F();
            sm.OnInt(65);
            Assert.AreEqual(sm.state_info(), "9");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F3" });

            sm = new BranchController();
            sm.F();
            sm.OnInt(115);
            Assert.AreEqual(sm.state_info(), "7");
            Assert.AreEqual(sm.tape, new List<string> { "-> $F1" });
        }




    }

}
